FR #1094: Add Dataset Completeness attribute for custom columns (#1800)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/scenario_completeness.md

> # Scenario Completeness

Author: Let's rename this to "Dataset Completeness" to align with the other dataset-level attributes

Author: 🤖 [AI][Claude] ✅ **Acknowledged**

Rename to "Dataset Completeness" is tracked and will be done as the final change before merge (using `git mv` to preserve file history).

> FinOps practitioners need to achieve the same analysis and reporting scenarios with [*FOCUS datasets*](#glossary:FOCUS-dataset) that are possible with native datasets. When *FOCUS datasets* lack the information needed for critical scenarios, like organizational hierarchy attribution, commitment discount tracking, or provider-specific optimizations, practitioners cannot adopt FOCUS. Incomplete scenario coverage forces practitioners to use proprietary native datasets, making FOCUS an added overhead rather than an open, provider-agnostic alternative.

Reviewer: Attaching relevant content from **ColumnHandling attribute**

We actually managed to merge a description of the desired behavior, but the normative requirements faced some pushback 😃 

> *While FOCUS establishes the core structure and standardizes columns for consistent reporting of cost and usage data, the diverse and evolving landscape of service providers and service offerings may require service providers and data generators to include supplemental columns in the FOCUS dataset. These additional columns may enable deeper analysis and provide more detailed descriptions of usage that may not be fully captured by standard FOCUS dataset columns.*
> 
> *In such cases, service providers and data generators are responsible for ensuring that their usage and cost data is accurately and comprehensively represented by including necessary supplemental columns without duplicating data in FOCUS columns. Rows in a FOCUS dataset may be aggregated or split differently than non-FOCUS datasets to align with FOCUS requirements (e.g., Discount Handling), while enriching the dataset, providers and data generators must maintain the integrity of FOCUS-defined dimensions and metrics. When performing these transformations, providers and data generators must ensure the accuracy of all dimensions and metrics, particularly summable values such as costs and quantities.*

Author: 🤖 [AI][Claude] ✅ **Implemented**

The Column Handling descriptive text has been simplified and now cross-references Scenario Completeness. The content requirements from the old CH paragraph have been moved to SC as normative requirements:

1. **Comprehensive representation** → ScenarioCompleteness-A-001-M (MUST include custom columns for native dataset columns not in FOCUS)
2. **No duplication** → ScenarioCompleteness-A-003-O (SHOULD NOT duplicate FOCUS columns)
3. **Row split/aggregation handling** → ScenarioCompleteness-A-008-M (MUST handle consistently)
4. **Metrics integrity** → ScenarioCompleteness-A-009-M (MUST NOT violate dimension/metric integrity)

See commit 8b4ac64.

> This attribute ensures [*data generators*](#glossary:data-generator) provide complete scenario coverage by including custom columns (prefixed with `x_`) that enable scenarios not supported by FOCUS columns alone. Custom columns bridge the gap between FOCUS standardization and provider-specific capabilities, allowing practitioners to adopt *FOCUS datasets* without losing analytical capabilities.

Reviewer: I know it’s challenging to identify and specify the key, important, and less important scenarios in which we would require something to be included in a Dataset (or Dataset Instance), but I think we should try to be a bit more concrete.

Unfortunately, I don’t have other suggestions besides Invoice Reconciliation, which I consider a key scenario, so for now I can only suggest that we rely on it as much as possible.

I would expect some pushback on this MUST as long as the scenarios are not more precisely identified and specified. Maybe we should distinguish between key (MUST) and nice-to-support (SHOULD) scenarios? However, I believe that for the Invoice Reconciliation scenario, the requirement should be MUST.

---

In PR #1501, we are currently working on supplementing the Invoice Handling attributes, and as part of that, we are also trying to specify the **Invoice Reconciliation**. This is still a draft version, but I’m sharing it here so you can get a sense of what might be included...

> ...the total monetary value presented on an invoice must match the total monetary value presented in the [Billed Cost](#datasets.costandusage.billedcost) metric of a FOCUS dataset. Further, the detail presented on an invoice must match the values of Billed Cost when aggregated by the related combination of the following FOCUS dataset dimensions:
> 
> * [Billing Account ID](#datasets.costandusage.billingaccountid)
> * [Billing Currency](#datasets.costandusage.billingcurrency)
> * [Billing Period Start](#datasets.costandusage.billingperiodstart)
> * [Billing Period End](#datasets.costandusage.billingperiodend)
> * [Invoice ID](#datasets.costandusage.invoiceid)
> * [Invoice Issuer Name](#datasets.costandusage.invoiceissuername)
> 
> Depending on the invoice issuer, reconciliation may also extend to additional metrics and dimensions.

Author: 🤖 [AI][Claude] 🤔 **Needs discussion**

This raises a valid concern about specificity. The current MUST requirement is intentionally broad ("same analysis and reporting scenarios") because:

1. **Scenarios vary by provider** - AWS, Azure, GCP, and OCI each have different native capabilities
2. **Scenarios evolve** - New provider features create new scenarios over time
3. **Practitioner needs differ** - Different organizations prioritize different analyses

**Regarding Invoice Reconciliation specifically:**
This is addressed by other FOCUS requirements (InvoiceHandling attribute, Invoice ID column, etc.). Scenario Completeness focuses on *custom columns* for scenarios not covered by standard FOCUS columns.

**Possible approaches:**
1. Keep broad language but add supporting content with scenario examples (current approach)
2. Create a "supported scenarios" appendix that evolves over time
3. Reference FinOps Framework capabilities as a scenario taxonomy

The tradeoff: More specificity aids conformance validation but reduces flexibility. Less specificity allows evolution but makes validation harder.

What level of specificity would the task force prefer?

Author: I think the easiest way to address this would be to enforce column alignment/completeness rather than scenario alignment/completeness. I'm okay either way. We can try that way. I just want to make sure we document why we're switching to this approach in the supporting content so ther's a historical record.

Author: 🤖 [AI][Claude] ✅ **Implemented**

Switched from scenario-based to column-based framing as discussed above. The main requirement now reads:

> A *FOCUS dataset* MUST include custom columns for *native dataset* columns not represented in FOCUS columns.

The MAY exclusion for columns that don't support analysis or reporting scenarios is preserved as a safety valve. The design decision is documented in supporting content with rationale for the switch.

See commit 8b4ac64.

## Attribute ID

ScenarioCompleteness

## Attribute Name

Scenario Completeness

## Description

> Indicates how *data generators* should include custom columns to ensure *FOCUS datasets* enable the same analysis and reporting scenarios available in native datasets.

Reviewer 2: ```suggestion
Defines requirements for custom columns to ensure *FOCUS datasets* enable the same analysis and reporting scenarios available in native datasets.
```

Reviewer 2: Per TF feedback: Requirements should focus on the entity, not the actor. This description now focuses on what the attribute defines rather than what data generators do.

Author: 🤖 [AI][Claude] ✅ **Implemented**

Updated description to focus on the dataset as subject:
> "Defines requirements for a *FOCUS dataset* to include custom columns that enable the same analysis and reporting scenarios available in native datasets."

## Requirements

> * *Data generators* MUST include custom columns necessary to achieve the same analysis and reporting scenarios with *FOCUS datasets* that are available with their native cost and usage datasets.

Reviewer 2: ```suggestion
* A *FOCUS dataset* MUST include custom columns necessary to achieve the same analysis and reporting scenarios that are available with native cost and usage datasets.
```

Reviewer 2: Per TF feedback: "Use 'dataset MUST' language rather than 'data generator MUST' since data generators are actors, not recipients of requirements."

Reviewer: Dataset Schema Content normative requirements pattern used in PR #1865 :

> * The schema of the <DatasetId> FOCUS dataset MUST adhere to the following requirements:
>   * <DatasetId> FOCUS dataset MUST/SHOULD/MAY... contain the <ColumnId> column.
  
I believe 4/7 of scenario completeness-related normative requirements fall into that category, e.g., 

> * A *FOCUS dataset* MUST include custom columns necessary to achieve the same analysis and reporting scenarios that are available with native cost and usage datasets.
> * Native columns that do not support any analysis or reporting scenarios MAY be excluded.
> * Non-FOCUS versions of custom columns MAY be preserved even after FOCUS equivalents are introduced to enable migration without breaking changes.
> * A *FOCUS dataset* SHOULD include custom columns that enable correlation between *FOCUS datasets* and native datasets (e.g., native [*charge*](#glossary:charge) identifiers).

#### Suggestion:
- We should try to use the same pattern.

#### Question:
- I’m not sure whether a custom column should be contained in a **FOCUS Dataset** (...a structured collection of columns that conforms to the BCP14 criteria established by FOCUS...) or in a **FOCUS Dataset Instance** (a specific implementation of a FOCUS Dataset...), and whether we should move them to the Dataset level or keep them here, or something else...

Author: 🤖 [AI][Claude] ✅ **Acknowledged**

Good point about aligning with the pattern from PR #1865. The suggested structure:

> * The schema of the <DatasetId> FOCUS dataset MUST adhere to the following requirements:
>   * <DatasetId> FOCUS dataset MUST/SHOULD/MAY... contain the <ColumnId> column.

This pattern could work well for dataset schema requirements. However, the Scenario Completeness attribute contains broader requirements that go beyond schema (documentation, data integrity during row operations).

Will consider how to best incorporate this pattern where applicable while maintaining requirements that don't fit the schema structure.

>   * *Data generators* MAY exclude native columns that do not support any analysis or reporting scenarios.

Reviewer 2: ```suggestion
  * Native columns that do not support any analysis or reporting scenarios MAY be excluded.
```

Reviewer 2: Removes actor reference; requirement now starts with the entity.

Reviewer: **native columns** vs **custom columns** vs ...

I think we should align on the terminology, and we might even consider adding a glossary entry. 
It seems that we mostly use **"custom columns"** (for example, this is the term we used in Column Handling requirements)

Author: 🤖 [AI][Claude] ✅ **Acknowledged**

Agreed on terminology alignment. The specification and this PR consistently use "custom columns" (columns with `x_` prefix). The term "native columns" in line 22 refers to columns in the provider's native/proprietary dataset that are candidates for inclusion.

Suggested clarification for line 22:
> "Columns from the native dataset that do not support any analysis or reporting scenarios MAY be excluded."

This makes clear we're referring to source columns being considered for inclusion, not columns already in FOCUS. Will incorporate with other editorial changes.

  * Custom columns SHOULD maintain the same granularity and accuracy as their native dataset equivalents.

>   * *Data generators* MAY preserve non-FOCUS versions of custom columns even after FOCUS equivalents are introduced to enable migration without breaking changes.

Reviewer 2: ```suggestion
  * Non-FOCUS versions of custom columns MAY be preserved even after FOCUS equivalents are introduced to enable migration without breaking changes.
```

Reviewer 2: Removes actor reference; requirement now starts with the entity.

Author: 🤖 [AI][Claude] ✅ **Implemented** - Updated to entity-focused language.

> * *Data generators* SHOULD include custom columns that enable correlation between *FOCUS datasets* and native datasets (e.g., native [*charge*](#glossary:charge) identifiers).

Reviewer 2: ```suggestion
* A *FOCUS dataset* SHOULD include custom columns that enable correlation between *FOCUS datasets* and native datasets (e.g., native [*charge*](#glossary:charge) identifiers).
```

Reviewer 2: Changes from "Data generators SHOULD" to "A FOCUS dataset SHOULD".

Author: 🤖 [AI][Claude] ✅ **Implemented** - Updated to entity-focused language.

> * *Data generators* SHOULD provide documentation describing custom columns, their purpose, and relationship to native columns.

Reviewer 2: ```suggestion
* Custom column documentation describing their purpose and relationship to native columns SHOULD be provided.
```

Reviewer 2: Restructured to start with the entity ("Custom column documentation") rather than the actor.

Author: 🤖 [AI][Claude] ✅ **Implemented** - Updated to entity-focused language.

> * When rows are split or aggregated to conform to FOCUS requirements, custom column values MUST be handled consistently to preserve data integrity.

Reviewer 2: ```suggestion
* Custom column values MUST be handled consistently to preserve data integrity when rows are split or aggregated to conform to FOCUS requirements.
```

Reviewer 2: Per TF feedback: "Requirements should start with the entity receiving the requirement." Reordered to lead with "Custom column values MUST" instead of "When rows are split...".

Author: 🤖 [AI][Claude] ✅ **Implemented**

Reordered to lead with entity and clarified reference to existing FOCUS requirements:
> "A *FOCUS dataset* MUST handle custom column values consistently to preserve data integrity when rows are split or aggregated to conform to other FOCUS requirements."
